File: go/serializer.go
package ason

import (
	"strconv"
	"strings"
	"unicode"
)

// Serialize converts a Value to its ASON string representation.
func Serialize(v *Value) string {
	if v == nil {
		return "null"
	}

	var sb strings.Builder
	serializeValue(&sb, v)
	return sb.String()
}

func serializeValue(sb *strings.Builder, v *Value) {
	if v == nil {
		sb.WriteString("null")
		return
	}

	switch v.typ {
	case TypeNull:
		sb.WriteString("null")
	case TypeBool:
		if v.boolVal {
			sb.WriteString("true")
		} else {
			sb.WriteString("false")
		}
	case TypeInteger:
		sb.WriteString(strconv.FormatInt(v.intVal, 10))
	case TypeFloat:
		sb.WriteString(strconv.FormatFloat(v.floatVal, 'g', -1, 64))
	case TypeString:
		serializeString(sb, v.strVal)
	case TypeArray:
		sb.WriteRune('[')
		for i, item := range v.arrVal {
			if i > 0 {
				sb.WriteRune(',')
			}
			serializeValue(sb, item)
		}
		sb.WriteRune(']')
	case TypeObject:
		sb.WriteRune('(')
		for i, key := range v.objKeys {
			if i > 0 {
				sb.WriteRune(',')
			}
			serializeValue(sb, v.objVals[key])
		}
		sb.WriteRune(')')
	}
}

func serializeString(sb *strings.Builder, s string) {
	// Check if quoting is needed
	needsQuote := false
	if len(s) == 0 {
		needsQuote = true
	} else {
		for _, r := range s {
			if r == '"' || r == '\\' || r == '(' || r == ')' ||
				r == '[' || r == ']' || r == '{' || r == '}' ||
				r == ',' || r == ':' || unicode.IsSpace(r) {
				needsQuote = true
				break
			}
		}
		// Check if it looks like a keyword or number
		if !needsQuote {
			if s == "null" || s == "true" || s == "false" {
				needsQuote = true
			} else if len(s) > 0 && (s[0] == '-' || s[0] == '+' || (s[0] >= '0' && s[0] <= '9')) {
				needsQuote = true
			}
		}
	}

	if !needsQuote {
		sb.WriteString(s)
		return
	}

	sb.WriteRune('"')
	for _, r := range s {
		switch r {
		case '"':
			sb.WriteString("\\\"")
		case '\\':
			sb.WriteString("\\\\")
		case '\n':
			sb.WriteString("\\n")
		case '\r':
			sb.WriteString("\\r")
		case '\t':
			sb.WriteString("\\t")
		default:
			if r < 0x20 {
				sb.WriteString("\\u")
				sb.WriteString(strconv.FormatInt(int64(r), 16))
			} else {
				sb.WriteRune(r)
			}
		}
	}
	sb.WriteRune('"')
}

// SerializeWithSchema serializes a value with its schema.
func SerializeWithSchema(v *Value) string {
	if v == nil {
		return "null"
	}
	if !v.IsObject() && !v.IsArray() {
		return Serialize(v)
	}

	var sb strings.Builder

	// Handle array of objects
	if v.IsArray() && v.Len() > 0 && v.Get(0).IsObject() {
		// Build schema from first object
		first := v.Get(0)
		sb.WriteRune('{')
		for i, key := range first.Keys() {
			if i > 0 {
				sb.WriteRune(',')
			}
			sb.WriteString(key)
			writeFieldSchema(&sb, first.Field(key))
		}
		sb.WriteRune('}')
		sb.WriteRune(':')

		// Serialize all objects
		for i := 0; i < v.Len(); i++ {
			if i > 0 {
				sb.WriteRune(',')
			}
			serializeObjectData(&sb, v.Get(i))
		}
		return sb.String()
	}

	// Single object
	if v.IsObject() {
		sb.WriteRune('{')
		for i, key := range v.Keys() {
			if i > 0 {
				sb.WriteRune(',')
			}
			sb.WriteString(key)
			writeFieldSchema(&sb, v.Field(key))
		}
		sb.WriteRune('}')
		sb.WriteRune(':')
		serializeObjectData(&sb, v)
		return sb.String()
	}

	return Serialize(v)
}

func writeFieldSchema(sb *strings.Builder, v *Value) {
	if v == nil {
		return
	}
	if v.IsArray() {
		sb.WriteString("[]")
		if v.Len() > 0 && v.Get(0).IsObject() {
			sb.WriteRune('{')
			first := v.Get(0)
			for i, key := range first.Keys() {
				if i > 0 {
					sb.WriteRune(',')
				}
				sb.WriteString(key)
				writeFieldSchema(sb, first.Field(key))
			}
			sb.WriteRune('}')
		}
	} else if v.IsObject() {
		sb.WriteRune('{')
		for i, key := range v.Keys() {
			if i > 0 {
				sb.WriteRune(',')
			}
			sb.WriteString(key)
			writeFieldSchema(sb, v.Field(key))
		}
		sb.WriteRune('}')
	}
}

func serializeObjectData(sb *strings.Builder, v *Value) {
	if v == nil || !v.IsObject() {
		serializeValue(sb, v)
		return
	}

	sb.WriteRune('(')
	for i, key := range v.Keys() {
		if i > 0 {
			sb.WriteRune(',')
		}
		field := v.Field(key)
		if field.IsObject() {
			serializeObjectData(sb, field)
		} else if field.IsArray() {
			serializeArrayData(sb, field)
		} else {
			serializeValue(sb, field)
		}
	}
	sb.WriteRune(')')
}

func serializeArrayData(sb *strings.Builder, v *Value) {
	if v == nil || !v.IsArray() {
		serializeValue(sb, v)
		return
	}

	sb.WriteRune('[')
	for i := 0; i < v.Len(); i++ {
		if i > 0 {
			sb.WriteRune(',')
		}
		item := v.Get(i)
		if item.IsObject() {
			serializeObjectData(sb, item)
		} else {
			serializeValue(sb, item)
		}
	}
	sb.WriteRune(']')
}

File: python/serializer.py
"""ASON Serializer."""

from typing import Any, Dict, List, Union

AsonValue = Union[None, bool, int, float, str, List[Any], Dict[str, Any]]


def serialize(obj: AsonValue) -> str:
    """Serialize a Python object to ASON format (data only)."""
    return _serialize_value(obj)


def serialize_with_schema(obj: AsonValue) -> str:
    """Serialize a Python object to ASON format with schema."""
    if obj is None:
        return "null"
    if not isinstance(obj, (dict, list)):
        return _serialize_value(obj)
    
    # Handle list of objects
    if isinstance(obj, list) and obj and isinstance(obj[0], dict):
        first = obj[0]
        schema = _build_schema(first)
        data_parts = [_serialize_object_data(item) for item in obj]
        return f"{schema}:{','.join(data_parts)}"
    
    # Single object
    if isinstance(obj, dict):
        schema = _build_schema(obj)
        data = _serialize_object_data(obj)
        return f"{schema}:{data}"

    return _serialize_value(obj)


def _serialize_value(v: AsonValue) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, int):
        return str(v)
    if isinstance(v, float):
        return str(v)
    if isinstance(v, str):
        return _serialize_string(v)
    if isinstance(v, list):
        items = [_serialize_value(item) for item in v]
        return f"[{','.join(items)}]"
    if isinstance(v, dict):
        items = [_serialize_value(val) for val in v.values()]
        return f"({','.join(items)})"
    return "null"


def _serialize_string(s: str) -> str:
    # Check if quoting is needed
    if not s:
        return '""'

    needs_quote = False
    for ch in s:
        if ch in '"\\()[]{},:' or ch.isspace():
            needs_quote = True
            break

    # Check for keywords and numbers
    if not needs_quote:
        if s in ("null", "true", "false"):
            needs_quote = True
        elif s and (s[0] in "+-" or s[0].isdigit()):
            needs_quote = True

    if not needs_quote:
        return s

    # Quote and escape
    result = ['"']
    for ch in s:
        if ch == '"':
            result.append('\\"')
        elif ch == '\\':
            result.append('\\\\')
        elif ch == '\n':
            result.append('\\n')
        elif ch == '\r':
            result.append('\\r')
        elif ch == '\t':
            result.append('\\t')
        else:
            result.append(ch)
    result.append('"')
    return ''.join(result)


def _build_schema(obj: Dict[str, Any]) -> str:
    parts = []
    for key, val in obj.items():
        if isinstance(val, list):
            if val and isinstance(val[0], dict):
                inner_schema = _build_schema(val[0])
                parts.append(f"{key}[{inner_schema}]")
            else:
                parts.append(f"{key}[]")
        elif isinstance(val, dict):
            inner_schema = _build_schema(val)
            parts.append(f"{key}{inner_schema}")
        else:
            parts.append(key)
    return "{" + ",".join(parts) + "}"


def _serialize_object_data(obj: Dict[str, Any]) -> str:
    parts = []
    for val in obj.values():
        if isinstance(val, dict):
            parts.append(_serialize_object_data(val))
        elif isinstance(val, list):
            parts.append(_serialize_array_data(val))
        else:
            parts.append(_serialize_value(val))
    return f"({','.join(parts)})"


def _serialize_array_data(arr: List[Any]) -> str:
    parts = []
    for item in arr:
        if isinstance(item, dict):
            parts.append(_serialize_object_data(item))
        else:
            parts.append(_serialize_value(item))
    return f"[{','.join(parts)}]"

